lib/ansible/plugins/list.py

```diff
@@ -126,10 +132,10 @@ def list_collection_plugins(ptype, collections, search_paths=None):
     for collection in collections.keys():
         if collection == 'ansible.builtin':
             # dirs from ansible install, but not configured paths
-            dirs.extend([d.path for d in ploader._get_paths_with_context() if d.internal])
+            dirs = [d.path for d in ploader._get_paths_with_context() if d.internal]
```

Comment:
with multiple collections this would overwrite previous ones, not add to them

Reply:
Listing `ansible.legacy` before this change would search the legacy paths twice - once for ansible.legacy, and again for ansible.builtin. This caused issues for differentiating between `ansible.builtin` and `ansible.legacy` (which is probably partly why the `do_replace_legacy` logic probably existed). So I separated the search paths per loop - now each collection searches only its own paths for plugins.